File: PythonAPI/carla/agents/navigation/constant_velocity_agent.py
# Copyright (c) # Copyright (c) 2018-2020 CVC.
#
# This work is licensed under the terms of the MIT license.
# For a copy, see <https://opensource.org/licenses/MIT>.

"""
This module implements an agent that roams around a track following random
waypoints and avoiding other vehicles. The agent also responds to traffic lights.
It can also make use of the global route planner to follow a specifed route
"""

import carla

from agents.navigation.basic_agent import BasicAgent
from agents.tools.misc import get_speed

class ConstantVelocityAgent(BasicAgent):
    """
    ConstantVelocityAgent implements an agent that navigates the scene at a fixed velocity.
    This agent will fail if asked to perform turns that are impossible are the desired speed.
    This includes lane changes. When a collision is detected, the constant velocity will stop,
    wait for a bit, and then start again.
    """

    def __init__(self, vehicle, target_speed=20, opt_dict={}):
        """
        Initialization the agent parameters, the local and the global planner.

            :param vehicle: actor to apply to agent logic onto
            :param target_speed: speed (in Km/h) at which the vehicle will move
            :param opt_dict: dictionary in case some of its parameters want to be changed.
                This also applies to parameters related to the LocalPlanner.
        """
        super(ConstantVelocityAgent, self).__init__(vehicle, target_speed, opt_dict=opt_dict)

        self._use_basic_behavior = False  # Whether or not to use the BasicAgent behavior when the constant velocity is down
        self._target_speed = target_speed / 3.6  # [m/s]
        self._current_speed = vehicle.get_velocity().length()  # [m/s]
        self._constant_velocity_stop_time = None

        self._restart_time = float('inf')  # Time after collision before the constant velocity behavior starts again

        if 'restart_time' in opt_dict:
            self._restart_time = opt_dict['restart_time']
        if 'use_basic_behavior' in opt_dict:
            self._use_basic_behavior = opt_dict['use_basic_behavior']
        if 'force_lane_following' in opt_dict:
            self._force_lane_following = opt_dict['force_lane_following']

        self.is_constant_velocity_active = True
        self._set_collision_sensor()
        self._set_constant_velocity(target_speed)

    def set_target_speed(self, speed):
        """Changes the target speed of the agent"""
        self._target_speed = speed / 3.6
        self._local_planner.set_speed(speed)

    def stop_constant_velocity(self):
        """Stops the constant velocity behavior"""
        self.is_constant_velocity_active = False
        self._vehicle.disable_constant_velocity()
        self._constant_velocity_stop_time = self._world.get_snapshot().timestamp.elapsed_seconds

    def restart_constant_velocity(self):
        """Public method to restart the constant velocity"""
        self.is_constant_velocity_active = True
        self._set_constant_velocity(self._target_speed)

    def _set_constant_velocity(self, speed):
        """Forces the agent to drive at the specified speed"""
        self._vehicle.enable_constant_velocity(carla.Vector3D(speed, 0, 0))

    def run_step(self):
        """Execute one step of navigation."""
        if not self.is_constant_velocity_active:
            if self._world.get_snapshot().timestamp.elapsed_seconds - self._constant_velocity_stop_time > self._restart_time:
                self.restart_constant_velocity()
                self.is_constant_velocity_active = True
            elif self._use_basic_behavior:
                return super(ConstantVelocityAgent, self).run_step()
            else:
                return carla.VehicleControl()

        hazard_detected = False

        # Retrieve all relevant actors
        actor_list = self._world.get_actors()
        vehicle_list = actor_list.filter("*vehicle*")
        lights_list = actor_list.filter("*traffic_light*")

        vehicle_speed = self._vehicle.get_velocity().length()

        max_vehicle_distance = self._base_vehicle_threshold + vehicle_speed
        affected_by_vehicle, adversary, _ = self._vehicle_obstacle_detected(vehicle_list, max_vehicle_distance)
        if affected_by_vehicle:
            vehicle_velocity = self._vehicle.get_velocity()
            hazard_speed = vehicle_velocity.dot(adversary.get_velocity()) / vehicle_velocity.length()
            hazard_detected = True

        # Check if the vehicle is affected by a red traffic light
        max_tlight_distance = self._base_tlight_threshold + 0.3 * vehicle_speed
        affected_by_tlight, _ = self._affected_by_traffic_light(lights_list, max_tlight_distance)
        if affected_by_tlight:
            hazard_speed = 0
            hazard_detected = True

        # The longitudinal PID is overwritten by the constant velocity but it is
        # still useful to apply it so that the vehicle isn't moving with static wheels
        control = self._local_planner.run_step()
        if hazard_detected:
            self._set_constant_velocity(hazard_speed)
        else:
            self._set_constant_velocity(self._target_speed)

        return control

    def _set_collision_sensor(self):
        blueprint = self._world.get_blueprint_library().find('sensor.other.collision')
        self._collision_sensor = self._world.spawn_actor(blueprint, carla.Transform(), attach_to=self._vehicle)
        self._collision_sensor.listen(lambda event: self.stop_constant_velocity())

File: carla/LibCarla/source/carla/agents/navigation/ConstantVelocityAgent.cpp
// Copyright (c) 2018-2020 CVC.
//
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT>.

#include "carla/agents/navigation/ConstantVelocityAgent.h"

#include "carla/agents/navigation/LocalPlanner.h"

#include "carla/client/Actor.h"
#include "carla/client/ActorBlueprint.h"
#include "carla/client/ActorList.h"
#include "carla/client/BlueprintLibrary.h"
#include "carla/client/Sensor.h"
#include "carla/client/TrafficLight.h"
#include "carla/client/Vehicle.h"
#include "carla/client/World.h"
#include "carla/client/WorldSnapshot.h"
#include "carla/client/Timestamp.h"
#include "carla/geom/Transform.h"
#include "carla/geom/Vector3D.h"
#include "carla/rpc/VehicleControl.h"

#include <utility>

namespace carla {
namespace agents {
namespace navigation {

ConstantVelocityAgent::ConstantVelocityAgent(SharedPtr<client::Vehicle> vehicle,
                                             float target_speed_kmh,
                                             const ConstantOptions &opts,
                                             SharedPtr<client::Map> map,
                                             SharedPtr<GlobalRoutePlanner> grp)
    : BasicAgent(std::move(vehicle), target_speed_kmh,
                 static_cast<const Options &>(opts), std::move(map), std::move(grp)),
      _use_basic_behavior(opts.use_basic_behavior),
      _target_speed_ms(target_speed_kmh / 3.6f),
      _restart_time(opts.restart_time),
      _is_constant_velocity_active(true),
      _constant_velocity_stop_time(0.0),
      _has_stop_timestamp(false),
      _collision_sensor(nullptr) {

  SetCollisionSensor();
  SetConstantVelocity(_target_speed_ms);
}

ConstantVelocityAgent::~ConstantVelocityAgent() {
  DestroySensor();
}

void ConstantVelocityAgent::SetTargetSpeed(float speed_kmh) {
  _target_speed_ms = speed_kmh / 3.6f;
  _local_planner->SetSpeed(speed_kmh);
}

void ConstantVelocityAgent::StopConstantVelocity() {
  _is_constant_velocity_active = false;
  _vehicle->DisableConstantVelocity();
  _constant_velocity_stop_time =
      _world->GetSnapshot().GetTimestamp().elapsed_seconds;
  _has_stop_timestamp = true;
}

void ConstantVelocityAgent::RestartConstantVelocity() {
  _is_constant_velocity_active = true;
  SetConstantVelocity(_target_speed_ms);
}

void ConstantVelocityAgent::SetConstantVelocity(float speed_ms) {
  _vehicle->EnableConstantVelocity(geom::Vector3D(speed_ms, 0.0f, 0.0f));
}

rpc::VehicleControl ConstantVelocityAgent::RunStep() {
  if (!_is_constant_velocity_active) {
    const double elapsed =
        _world->GetSnapshot().GetTimestamp().elapsed_seconds;
    if (_has_stop_timestamp &&
        (elapsed - _constant_velocity_stop_time) >
            static_cast<double>(_restart_time)) {
      RestartConstantVelocity();
      _is_constant_velocity_active = true;
    } else if (_use_basic_behavior) {
      return BasicAgent::RunStep();
    } else {
      return rpc::VehicleControl{};
    }
  }

  bool hazard_detected = false;
  float hazard_speed = 0.0f;

  auto actor_list   = _world->GetActors();
  auto vehicle_list = actor_list->Filter("*vehicle*");
  auto lights_list  = actor_list->Filter("*traffic_light*");

  const float vehicle_speed = _vehicle->GetVelocity().Length();

  // Vehicle obstacle check.
  const float max_vehicle_distance = _base_vehicle_threshold + vehicle_speed;
  auto v_obs = VehicleObstacleDetected(vehicle_list, max_vehicle_distance);
  if (v_obs.affected && v_obs.actor) {
    const auto vehicle_velocity = _vehicle->GetVelocity();
    const float v_len = vehicle_velocity.Length();
    if (v_len == 0.0f) {
      hazard_speed = 0.0f;
    } else {
      const auto adv_velocity = v_obs.actor->GetVelocity();
      const float dot = vehicle_velocity.x * adv_velocity.x +
                        vehicle_velocity.y * adv_velocity.y +
                        vehicle_velocity.z * adv_velocity.z;
      hazard_speed = dot / v_len;
    }
    hazard_detected = true;
  }

  // Red-light check (uses 0.3 * speed factor instead of _speed_ratio — match
  // the Python source verbatim).
  const float max_tlight_distance = _base_tlight_threshold + 0.3f * vehicle_speed;
  auto tl_pair = AffectedByTrafficLight(lights_list, max_tlight_distance);
  if (tl_pair.first) {
    hazard_speed = 0.0f;
    hazard_detected = true;
  }

  // Run the local planner so the wheels keep spinning even though the actual
  // forward motion comes from EnableConstantVelocity.
  rpc::VehicleControl control = _local_planner->RunStep();
  if (hazard_detected) {
    SetConstantVelocity(hazard_speed);
  } else {
    SetConstantVelocity(_target_speed_ms);
  }
  return control;
}

void ConstantVelocityAgent::SetCollisionSensor() {
  auto blueprint_lib = _world->GetBlueprintLibrary();
  const auto *blueprint = blueprint_lib->Find("sensor.other.collision");
  if (!blueprint) {
    return;
  }

  auto actor = _world->SpawnActor(*blueprint, geom::Transform{}, _vehicle.get());
  _collision_sensor = std::dynamic_pointer_cast<client::Sensor>(actor);
  if (!_collision_sensor) {
    return;
  }

  // Capture `this` raw — the agent owns the sensor and tears it down in the
  // destructor, so the callback can never fire on a destroyed agent.
  _collision_sensor->Listen(
      [this](SharedPtr<sensor::SensorData>) {
        this->StopConstantVelocity();
      });
}

void ConstantVelocityAgent::DestroySensor() {
  if (_collision_sensor) {
    if (_collision_sensor->IsListening()) {
      _collision_sensor->Stop();
    }
    _collision_sensor->Destroy();
    _collision_sensor.reset();
  }
}

} // namespace navigation
} // namespace agents
} // namespace carla
